### PR TITLE
feat(inference): support generation using trainer

### PR DESCRIPTION
**Pull request by KimmiShi**
## Motivation

Support generation using trainer/engine, required in RLHF.

## Modification

- inference.py: support trainer/engine generation. refact code in getting inference params.
- scheduler: support more input arguments like scalar and inference params.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.

### CODE REVIEW

File: internlm/apis/inference.py [hunk 1]
```diff
@@ -964,3 +1000,123 @@ def top_k_top_p_filtering(logits, top_k=0, top_p=1.0, filter_value=-float("Inf")
...
+    Args:
+        batch (List[Dict]):  a list of prompts
+        pad_token_id (int, optional): Defaults to 0.
+        return_dict (bool, optional): Defaults to False.
+
+    Returns:
+        Union[Dict, torch.Tensor]: input_ids or dict(input_ids=input_ids)
+    """
+    input_ids = []
+    max_length = max([len(w["input_ids"] if isinstance(w, Dict) else w) for w in batch])
```
Comment (gaoyang07) [inline internlm/apis/inference.py:1083]: 这里需要更新下，这里传进来的batch并未考虑其本身就是一个str encode之后的list的情况（对应上面line 1054这个情况

File: internlm/apis/inference.py [hunk 2]
```diff
@@ -726,79 +736,105 @@ def _beam_search_generate(
...
-            # (bsz, 2*num_beams)
-            next_scores, ids = torch.topk(_scores, 2 * num_beams, dim=1, largest=True, sorted=True)
-            # (batch_size, 2*num_beams)
-            from_which_beam = torch.floor(ids.float() / vocab_size).long()
-            next_tokens = ids % vocab_size  # (batch_size, 2*num_beams)
+            next_tokens = tokens.new_zeros([batch_size, 2 * num_beams])
+            next_scores = torch.zeros([batch_size, 2 * num_beams], dtype=torch.float32, device=next_tokens.device)
+            from_which_beam = torch.zeros([batch_size, 2 * num_beams], dtype=torch.int64, device=next_tokens.device)
+
+        # import pdb;pdb.set_trace()
```
Comment (gaoyang07) [inline internlm/apis/inference.py]: 删掉debug代码～

File: internlm/apis/inference.py (hunk 1 above)
Comment (gaoyang07) [reply internlm/apis/inference.py:1083]: 如果一定传一个`[[content]]`这样的一个结构，那可以去掉line 1054的逻辑

**Review by gaoyang07 — COMMENTED** [review]
以及补充一个测例，某条数据拼起来做batch generation的结果需要与这条数据单独generation的结果一致（在推理参数一致的情况下）

File: internlm/apis/inference.py (hunk 1 above)
Comment (KimmiShi) [reply internlm/apis/inference.py:1083]: ok，更新为如下逻辑：
1. batch_tokenize_process_fn在只输入一个str时候报错。
2. pad_input_ids 加入了检查。

**Comment by KimmiShi** [comment]
> 以及补充一个测例，某条数据拼起来做batch generation的结果需要与这条数据单独generation的结果一致（在推理参数一致的情况下）

已添加在test_generate.py

**Review by gaoyang07 — APPROVED** [review]
